metaserver/src/main/scala/scala/meta/languageserver/providers/DocumentFormattingProvider.scala

```diff
@@ -1,52 +1,91 @@
 package scala.meta.languageserver.providers
 
 import java.nio.file.Files
-import scala.meta.languageserver.Formatter
 import scala.meta.languageserver.Configuration
+import scala.meta.languageserver.Configuration.Scalafmt
+import scala.meta.languageserver.Formatter
+import scala.meta.languageserver.MonixEnrichments._
+import scala.util.control.NonFatal
 import com.typesafe.scalalogging.LazyLogging
+import langserver.core.Notifications
 import langserver.messages.DocumentFormattingResult
+import langserver.types.MessageType
 import langserver.types.Position
 import langserver.types.Range
 import langserver.types.TextEdit
+import monix.eval.Task
+import monix.execution.Scheduler
+import monix.reactive.Observable
 import org.langmeta.inputs.Input
 import org.langmeta.io.AbsolutePath
-import monix.eval.Task
 
 class DocumentFormattingProvider(
-    configuration: Task[Configuration],
-    cwd: AbsolutePath
-) extends LazyLogging {
+    configuration: Observable[Configuration],
+    cwd: AbsolutePath,
+    notifications: Notifications
+)(implicit s: Scheduler)
+    extends LazyLogging {
 
-  private lazy val scalafmt = Formatter.classloadScalafmt("1.3.0")
+  private val formatter: () => Either[String, Formatter] =
+    configuration
+      .focus(_.scalafmt.version)
+      .map[Either[String, Formatter]] { version =>
+        try {
+          // TODO(olafur) convert Jars.fetch to use monix.Task to avoid blocking
+          Right(Formatter.classloadScalafmt(version))
+        } catch {
+          case NonFatal(e) =>
+            val msg =
+              s"Unable to install scalafmt version $version, cause: ${e.getMessage}"
+            Left(msg)
+        }
+      }
+      .toFunction0()
 
-  private def formatterFromConfiguration: Task[Formatter] =
-    configuration.map { conf =>
-      if (conf.scalafmt.enabled) scalafmt
-      else Formatter.noop
-    }
+  private val config: () => Either[String, Option[AbsolutePath]] =
+    configuration
+      .focus(_.scalafmt.confPath)
+      .map[Either[String, Option[AbsolutePath]]] {
+        case None =>
+          val default = cwd.resolve(Scalafmt.DefaultConf)
+          if (Files.isRegularFile(default.toNIO)) Right(Some(default))
```

Comment:
This can probably be a shared utility function:

```scala
def pathIfExists(path: AbsolutePath): Option[AbsolutePath] =
  Option(path).filter(p => Files.isRegularFile(p.toNIO))
```